### PR TITLE
Support cloning via VCS tokens

### LINKED ISSUE #81: Artemis Token API & Authorization to GIT via SSH
Tracking https://github.com/ls1intum/Artemis/pull/8929 and https://github.com/ls1intum/Artemis/pull/9182

### PR DESCRIPTION
**Pull request by Feuermagier**
Closes #81 

Support repository cloning using the user-wide VCS token. The implementation uses an existing token, if present, or generates a new one with a default lifetime of five days (which is subject to discussion, I just picked some random value). SSH cloning is still supported.

I've also refactored the cloning code a bit by introducing a proper strategy pattern.

### CODE REVIEW

File: src/main/java/edu/kit/kastel/sdq/artemis4j/grading/git/VCSTokenCloningStrategy.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        String token;
+        if (tokenOverride != null) {
+            token = tokenOverride;
+        } else if (assessor.getGitToken().isPresent()) {
+            token = assessor.getGitToken().get();
+        } else {
+            // The user has not set a token, so create one for him with a default expiration time
+            log.info("Creating new VCS access token");
+            var expiryDate = ZonedDateTime.now().plusDays(5);
```
Comment (dfuchss) [inline src/main/java/edu/kit/kastel/sdq/artemis4j/grading/git/VCSTokenCloningStrategy.java:35]: Please check how long the max time currently is (hard coded in artemis). I'd create a token at least for some weeks :)

File: src/main/java/edu/kit/kastel/sdq/artemis4j/grading/git/VCSTokenCloningStrategy.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    private final CredentialsProvider credentialsProvider;
+
+    public VCSTokenCloningStrategy(String tokenOverride, ArtemisConnection connection) throws ArtemisNetworkException {
+        var assessor = connection.getAssessor();
+
+        String token;
+        if (tokenOverride != null) {
+            token = tokenOverride;
+        } else if (assessor.getGitToken().isPresent()) {
+            token = assessor.getGitToken().get();
```
Comment (dfuchss) [inline src/main/java/edu/kit/kastel/sdq/artemis4j/grading/git/VCSTokenCloningStrategy.java:31]: Also check that the token does not getting invalid in the next X (e.g., 2) days. If so, regenerate the token.

**Review by dfuchss — CHANGES_REQUESTED** [review]
Code looks good to me. Just some suggestions regarding the token handling

File: src/main/java/edu/kit/kastel/sdq/artemis4j/grading/git/VCSTokenCloningStrategy.java (hunk 1 above)
Comment (Feuermagier) [reply src/main/java/edu/kit/kastel/sdq/artemis4j/grading/git/VCSTokenCloningStrategy.java:35]: The max time is one year (https://github.com/ls1intum/Artemis/blob/921b1884fb1a3028512e61023f254f1365fec14b/src/main/java/de/tum/in/www1/artemis/web/rest/AccountResource.java#L175), but that feels way too long to me. What about 6 months? That ensures that tokens generated at the start of a semester are invalidated at its end.

Comment (dfuchss) [reply src/main/java/edu/kit/kastel/sdq/artemis4j/grading/git/VCSTokenCloningStrategy.java:35]: Fine for me :)

**Review by Luro02 — APPROVED** [review]

**Review by dfuchss — APPROVED** [review]

**Pull request merged**